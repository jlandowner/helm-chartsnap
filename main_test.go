package main

import (
	"bytes"
	"os"
	"testing"

	"github.com/fatih/color"
	"github.com/jlandowner/helm-chartsnap/pkg/api/v1alpha1"
	"github.com/jlandowner/helm-chartsnap/pkg/charts"
	. "github.com/jlandowner/helm-chartsnap/pkg/snap/gomega"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

func TestMain(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Main Suite")
}

var _ = Describe("overrideSnapshotConfig", func() {
	type testCase struct {
		name     string
		opt      *option
		cfg      *v1alpha1.SnapshotConfig
		expected *v1alpha1.SnapshotConfig
	}

	DescribeTable("overriding snapshot config",
		func(tc testCase) {
			tc.opt.overrideSnapshotConfig(tc.cfg)

			Expect(tc.cfg.DynamicFields).To(Equal(tc.expected.DynamicFields))
			Expect(tc.cfg.SnapshotFileExt).To(Equal(tc.expected.SnapshotFileExt))
			Expect(tc.cfg.SnapshotVersion).To(Equal(tc.expected.SnapshotVersion))
		},
		Entry("when SnapshotFileExt is set in option", testCase{
			name: "SnapshotFileExt is set",
			opt: &option{
				SnapshotFileExt: "yaml",
			},
			cfg: &v1alpha1.SnapshotConfig{},
			expected: &v1alpha1.SnapshotConfig{
				SnapshotFileExt: "yaml",
			},
		}),
		Entry("when SnapshotVersion is set in option", testCase{
			name: "SnapshotVersion is set",
			opt: &option{
				SnapshotVersion: "v2",
			},
			cfg: &v1alpha1.SnapshotConfig{},
			expected: &v1alpha1.SnapshotConfig{
				SnapshotVersion: "v2",
			},
		}),
		Entry("when SnapshotFileExt is set in config", testCase{
			name: "SnapshotFileExt is set",
			opt:  &option{},
			cfg: &v1alpha1.SnapshotConfig{
				SnapshotFileExt: "yaml",
			},
			expected: &v1alpha1.SnapshotConfig{
				SnapshotFileExt: "yaml",
			},
		}),
		Entry("when SnapshotVersion is set in config", testCase{
			name: "SnapshotVersion is set",
			opt:  &option{},
			cfg: &v1alpha1.SnapshotConfig{
				SnapshotVersion: "v2",
			},
			expected: &v1alpha1.SnapshotConfig{
				SnapshotVersion: "v2",
			},
		}),
		Entry("when LegacySnapshot is true", testCase{
			name: "LegacySnapshot is true",
			opt: &option{
				LegacySnapshot: true,
			},
			cfg: &v1alpha1.SnapshotConfig{},
			expected: &v1alpha1.SnapshotConfig{
				SnapshotVersion: charts.SnapshotVersionV1,
			},
		}),
		Entry("when both SnapshotFileExt and SnapshotVersion are set", testCase{
			name: "both SnapshotFileExt and SnapshotVersion are set",
			opt: &option{
				SnapshotFileExt: "yaml",
				SnapshotVersion: "v3",
			},
			cfg: &v1alpha1.SnapshotConfig{
				SnapshotFileExt: "yml",
				SnapshotVersion: "v2",
			},
			expected: &v1alpha1.SnapshotConfig{
				SnapshotFileExt: "yaml",
				SnapshotVersion: "v3",
			},
		}),
	)
})

var _ = Describe("rootCmd", func() {
	BeforeEach(func() {
		initRootCmd()
	})
	Context("success", func() {
		Context("snapshot local chart with single values file", func() {
			It("should pass", func() {
				var output bytes.Buffer
				color.Output = &output
				DeferCleanup(func() {
					color.Output = os.Stdout
				})
				rootCmd.SetArgs([]string{"--chart", "example/app1", "-f", "example/app1/test_latest/test_ingress_enabled.yaml", "--namespace", "default"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
				Ω(output.String()).To(MatchSnapShot())
			})
		})

		Context("snapshot local chart with values directory", func() {
			It("should pass", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app1", "-f", "example/app1/test_latest/", "--namespace", "default"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
			})
		})

		Context("snapshot OCI chart", func() {
			It("should pass", func() {
				rootCmd.SetArgs([]string{"--chart", "oci://ghcr.io/nginxinc/charts/nginx-gateway-fabric", "-f", "example/remote/nginx-gateway-fabric.values.yaml", "--", "--namespace", "nginx-gateway"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
			})
		})

		Context("snapshot remote chart 1", func() {
			It("should pass", func() {
				rootCmd.SetArgs([]string{"--chart", "cilium", "-f", "example/remote/cilium.values.yaml", "--", "--namespace", "kube-system", "--repo", "https://helm.cilium.io"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
			})
		})

		Context("snapshot remote chart 2", func() {
			It("should pass", func() {
				rootCmd.SetArgs([]string{"--chart", "ingress-nginx", "-f", "example/remote/ingress-nginx.values.yaml", "--", "--namespace", "ingress-nginx", "--repo", "https://kubernetes.github.io/ingress-nginx", "--skip-tests"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
			})
		})

		Context("snapshot empty chart", func() {
			It("should pass", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app2", "--namespace", "default"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
			})
		})

		Context("snapshot helm error", func() {
			It("should pass", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app3", "--namespace", "default"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
			})
		})

		Context("snapshot empty chart with no config file", func() {
			It("should pass", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app2", "--namespace", "default", "--config-file", "notfound"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
			})
		})

		Context("snapshot empty chart with debug mode", func() {
			It("should pass", func() {
				os.Setenv("HELM_DEBUG", "true")
				defer os.Unsetenv("HELM_DEBUG")
				rootCmd.SetArgs([]string{"--chart", "example/app2", "--namespace", "default"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
			})
		})

		Context("env FORCE_COLOR is enabled", func() {
			It("should force a colorized output", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app1", "-f", "example/app1/test_latest/test_ingress_enabled.yaml", "--namespace", "default"})
				var output bytes.Buffer
				color.Output = &output
				os.Setenv("FORCE_COLOR", "1")
				DeferCleanup(func() {
					color.Output = os.Stdout
					color.NoColor = true
					os.Unsetenv("FORCE_COLOR")
				})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
				Ω(output.String()).To(MatchSnapShot())
			})
		})

		Context("snapshot file ext is yaml", func() {
			It("should pass", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app3", "--namespace", "default", "--snapshot-file-ext", "yaml"})
				err := rootCmd.Execute()
				Expect(err).ShouldNot(HaveOccurred())
			})
		})
	})

	Context("fail", func() {
		Context("including dynamic outputs", func() {
			It("should fail", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app1", "--namespace", "default"})
				err := rootCmd.Execute()
				Expect(err).To(HaveOccurred())
				Ω(err.Error()).To(MatchSnapShot())
			})
		})

		Context("snapshot is different", func() {
			It("should fail", func() {
				var output bytes.Buffer
				color.Output = &output
				DeferCleanup(func() {
					color.Output = os.Stdout
				})
				rootCmd.SetArgs([]string{"--chart", "example/app1", "--namespace", "default", "-f", "example/app1/testfail/test_ingress_enabled.yaml"})
				err := rootCmd.Execute()
				Expect(err).To(HaveOccurred())
				Ω(err.Error()).To(MatchSnapShot())
				Ω(output.String()).To(MatchSnapShot())
			})
		})

		Context("values directory contains not matched snapshots", func() {
			It("should fail", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app1", "--namespace", "default", "-f", "example/app1/testfail"})
				err := rootCmd.Execute()
				Expect(err).To(HaveOccurred())
				Ω(err.Error()).To(MatchSnapShot())
			})
		})

		Context("values file not found", func() {
			It("should fail", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app1", "-f", "example/app1/test_latest/notfound.yaml", "--namespace", "default"})
				err := rootCmd.Execute()
				Expect(err).To(HaveOccurred())
				Ω(err.Error()).To(MatchSnapShot())
			})
		})

		Context("snapshot helm error with --fail-helm-error", func() {
			It("should fail", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app3", "--namespace", "default", "--fail-helm-error"})
				err := rootCmd.Execute()
				Expect(err).To(HaveOccurred())
				Ω(err.Error()).To(MatchSnapShot())
			})
		})

		Context("required flag is not set", func() {
			It("should fail", func() {
				rootCmd.SetArgs([]string{})
				err := rootCmd.Execute()
				Expect(err).To(HaveOccurred())
				Ω(err.Error()).To(MatchSnapShot())
			})
		})

		Context("invalid flag", func() {
			It("should fail", func() {
				rootCmd.SetArgs([]string{"--chart", "example/app1", "-f", "example/app1/test_latest/test_ingress_enabled.yaml", "--namespace", "default", "--invalid"})
				err := rootCmd.Execute()
				Expect(err).To(HaveOccurred())
				Ω(err.Error()).To(MatchSnapShot())
			})
		})
	})

	Context("--help", func() {
		It("should show help", func() {
			rootCmd.SetArgs([]string{"--help"})
			help := rootCmd.UsageString()
			Ω(help).To(MatchSnapShot())
		})
	})
})
