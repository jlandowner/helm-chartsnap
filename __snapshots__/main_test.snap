['rootCmd --help should show help 1']
SnapShot = """
Usage:
  chartsnap -c CHART [flags]

Examples:

  # Snapshot with default values:
  chartsnap -c YOUR_CHART
  
  # Update snapshot files:
  chartsnap -c YOUR_CHART -u

  # Snapshot with test case values:
  chartsnap -c YOUR_CHART -f YOUR_TEST_VALUES_FILE
  
  # Snapshot all test cases:
  chartsnap -c YOUR_CHART -f YOUR_TEST_VALUES_FILES_DIRECTOY
  
  # Set additional args or flags for the 'helm template' command:
  chartsnap -c YOUR_CHART -f YOUR_TEST_VALUES_FILE -- --skip-tests

  # Snapshot remote chart in Helm repository:
  chartsnap -c CHART_NAME -f YOUR_VALUES_FILE -- --repo HELM_REPO_URL

  # Snapshot ingress-nginx (https://kubernetes.github.io/ingress-nginx/) helm chart for a specific version with your value file:
  chartsnap -c ingress-nginx -f YOUR_VALUES_FILE -- --repo https://kubernetes.github.io/ingress-nginx --namespace kube-system --version 4.8.3

  # Snapshot cilium (https://cilium.io) helm chart with default value and set flags:
  chartsnap -c cilium -- --repo https://helm.cilium.io --namespace kube-system --set hubble.relay.enabled=true --set hubble.ui.enabled=true

  # Snapshot charts in OCI registry
  chartsnap -c oci://ghcr.io/nginxinc/charts/nginx-gateway-fabric -n nginx-gateway

  # Output with no colors:
  NO_COLOR=1 chartsnap -c YOUR_CHART

Flags:
  -c, --chart string               path to the chart directory. this flag is passed to 'helm template RELEASE_NAME CHART --values VALUES' as 'CHART'
      --config-file string         config file name or path, which defines snapshot behavior e.g. dynamic fields (default \".chartsnap.yaml\")
  -N, --ctx-lines int              number of lines to show in diff output. 0 for full output (default 3)
      --debug                      debug mode
      --fail-helm-error            fail if 'helm template' command failed
      --failfast                   fail once any test case failed
  -n, --namespace string           namespace. this flag is passed to 'helm template RELEASE_NAME CHART --values VALUES --namespace NAMESPACE' as 'NAMESPACE' (default \"default\")
  -o, --output-dir string          directory which is __snapshot__ directory is created. (default: values file directory if --values is set; chart directory if chart is local; else current directory)
      --parallelism int            test concurrency if taking multiple snapshots for a test value file directory. default is unlimited (default -1)
      --release-name string        release name. this flag is passed to 'helm template RELEASE_NAME CHART --values VALUES' as 'RELEASE_NAME' (default \"chartsnap\")
      --snapshot-file-ext string   snapshot file extension. default is \".snap\" and if set \"yaml\", \".snap.yaml\" is used
      --snapshot-version string    use a specific snapshot format version. v1, v2, v3 are supported. (default: latest)
  -u, --update-snapshot            update snapshot mode
  -f, --values string              path to a test values file or directory. if the directory is set, all test files are tested. if empty, default values are used. this flag is passed to 'helm template RELEASE_NAME CHART --values VALUES' as 'VALUES'
"""

['rootCmd fail including dynamic outputs should fail 1']
SnapShot = 'snapshot does not match chart=example/app1 values='

['rootCmd fail invalid flag should fail 1']
SnapShot = 'unknown flag: --invalid'

['rootCmd fail required flag is not set should fail 1']
SnapShot = 'required flag(s) "chart" not set'

['rootCmd fail snapshot helm error with --fail-helm-error should fail 1']
SnapShot = """
failed to get snapshot chart=example/app3 values=: 'helm template' command failed: exit status 1: Error: execution error at (app3/templates/secret.yaml:8:13): apiKey is required

Use --debug flag to render out invalid YAML
"""

['rootCmd fail snapshot is different should fail 1']
SnapShot = 'snapshot does not match chart=example/app1 values=example/app1/testfail/test_ingress_enabled.yaml'

['rootCmd fail snapshot is different should fail 2']
SnapShot = """
 RUNS  Snapshot testing chart=example/app1 values=example/app1/testfail/test_ingress_enabled.yaml
 FAIL  Snapshot does not match chart=example/app1 values=example/app1/testfail/test_ingress_enabled.yaml snapshot_version=v2
"""

['rootCmd fail values directory contains not matched snapshots should fail 1']
SnapShot = 'snapshot does not match chart=example/app1 values=example/app1/testfail/test_ingress_enabled.yaml'

['rootCmd fail values file not found should fail 1']
SnapShot = """
values file 'example/app1/test_latest/notfound.yaml' not found"""

['rootCmd success env FORCE_COLOR is enabled should force a colorized output 1']
SnapShot = """
\u001B[37;44m RUNS \u001B[0m\u001B[0m Snapshot testing chart=example/app1 values=example/app1/test_latest/test_ingress_enabled.yaml
\u001B[0m\u001B[37;42m PASS \u001B[0m\u001B[32m Snapshot matched chart=example/app1 values=example/app1/test_latest/test_ingress_enabled.yaml snapshot_version=v3
\u001B[0m\u001B[37;42m PASS \u001B[0m\u001B[32m All snapshots matched
\u001B[0m"""

['rootCmd success snapshot local chart with single values file should pass 1']
SnapShot = """
 RUNS  Snapshot testing chart=example/app1 values=example/app1/test_latest/test_ingress_enabled.yaml
 PASS  Snapshot matched chart=example/app1 values=example/app1/test_latest/test_ingress_enabled.yaml snapshot_version=v3
 PASS  All snapshots matched
"""
